[Snippets][Doc] Updated the design guide (#22258)
openvinotoolkit/openvino: src/common/snippets/docs/snippets_design_guide.md

Note that if a `PortDescriptor` is required, you can obtain it directly (without getting `Expression` first) by calling `ExpressionPort->get_descriptor_ptr()`, so the `ExpressionPort` will snag the right `PortDescriptor` from the `Expression` for you.

Concluding this section, it's worth mentioning that the `LinearIR` currently provides several debug feature `debug_print()` and serialization. 
The first one prints input and output `PortConnectors` and `PortDescriptors` for every `Expression` to stderr. 
The second one allows users to serialize the `LIR` in two views: as control flow graph of `LinearIR` using the pass `SerializeControlFlow` and 
as data flow graph of `LinearIR` using the pass `SerializeDataFlow` (control flow operations (e.g. `LoopBegin`/`LoopEnd`) are not serialized).
The both serializations are saved in `xml` OpenVINO graph format, where a lot of useful parameters are displayed for every `Expression`.

> Also the `LinearIR` provides an interface for work with new `Expression` which encapsulate all the technical details inside.

Reviewer: ```suggestion
Also the `LinearIR` provides an interface for work with new `Expression` which encapsulates all the technical details inside.
```

> Please see [perf_count.md](./debug_capabilities/perf_count.md) and [snippets_segfault_detector.md](./debug_capabilities/snippets_segfault_detector.md) for more info regarding the performance counters and segfault detector. 

Reviewer 2: Since we now have https://github.com/openvinotoolkit/openvino/tree/master/src/common/snippets/docs/debug_capabilities it worth to move debug related info there.

as data flow graph of `LinearIR` using the pass `SerializeDataFlow` (control flow operations (e.g. `LoopBegin`/`LoopEnd`) are not serialized).
The both serializations are saved in `xml` OpenVINO graph format, where a lot of useful parameters are displayed for every `Expression`.
Please see [perf_count.md](./debug_capabilities/perf_count.md) and [snippets_segfault_detector.md](./debug_capabilities/snippets_segfault_detector.md) for more info regarding the performance counters and segfault detector. 

Also the `LinearIR` provides an interface for work with new `Expression` which encapsulates all the technical details inside.
The method `insert_node(...)` creates new `Expression` based on the passed `ov::Node` and inserts in the specified place in the `LinearIR`
by performing all the necessary actions for full integration into the `LinearIR` (connection with parents and consumers and updates of the corresponding `LoopInfo`).
This helper is used in several control flow transformations `InsertSomething`, for example, in the `InsertLoadStore` pass inside [insert_load_store.cpp](../src/lowered/pass/insert_load_store.cpp).

> The method `replace_with_node(...)` creates new `Expression` based on the passed `ov::Node` and replaces the existing `Expressions` in the `LinearIR`

Author: Everything is done. Thanks!

3. `InsertBuffers` - analyzes `LoopInfo` and `Expression` semantics, inserts `snippets::op::Buffer`. `Buffer` is an operation that represents a memory buffer needed to save some intermediate results.
4. `InsertLoadStore` - inserts explicit memory access operations like `Load` and `Store` (both from `snippets::op`). 
These are needed, so appropriate instructions will be emitted during the code generation stage to move data between memory and vector registers.
5. `InitLoops` - initialize data pointer shift parameters (pointer increments and finalization offsets for each loop port) in `LoopInfo`.
6. `InsertLoops` - inserts explicit `LoopBegin` and `LoopEnd` (`snippets::op`) operations based on the acquired `LoopInfo`. 
Again, the explicit operations are needed to emit appropriate instructions later.

> 7. `InsertBroadcastMove` insert `MoveBroadcast` before `Expressions` with several inputs since a special broadcasting instruction needs to be 

Reviewer 3: please remove linebreak

Author: Sure, done. Thanks!

generated to broadcast a single value to fill the whole vector register.
8. `LoadMoveBroadcastToBroadcastLoad` fuse `Load->MoveBroadcast` sequences into a single `BroadcastLoad` expressions.
9. `AllocateBuffers` is responsible for a safe `Buffer` data pointer increments and common memory size calculation. For more details refer please to the end of this section.
10. `CleanRepeatedDataPointerShifts` - eliminates redundant pointer increments from the loops.
11. `PropagateLayout` - propagates data layouts to `Parameters` and `Results` (actually to corresponding `Expressions`), so `Kernel` will be able to calculate appropriate data offsets for every iteration of an external parallel loop.

As mentioned above the `op::Buffer` operations are managed by the pass `AllocateBuffers`.

> Before describing the algorithm, it is necessary to briefly consider the structure of `Buffer`:

Reviewer 3: It looks like L609 and L610 are meant to be on separate lines. It would look much better if we put `*` before `All` and `Each` and make them a bullet list. Then `The algorithm supports` will start a new block of text

Author: Good idea. Thanks!

all `Buffers` represent `Buffer scratchpad` together (a common memory that is needed for intermediate results storing);
each `Buffer` has an `offset` relative to the common data pointer (pointer of `Buffer scratchpad`) and `ID` (the `Buffers` with the same `ID` have the same assigned register).

> The algorithm supports two modes: optimized (the algorithm calculates minimal memory size and minimal unique `ID` count required to handle all the buffers) and

Reviewer 3: It would be more readable if we split it in several shorter sentences:
```suggestion
The algorithm supports two modes: optimized and non-optimized. The optimized one calculates minimal memory size and minimal unique `ID` count required to handle all the buffers. The non-optimized version assigns each buffer a unique...
```

Author: Thanks for the suggestion, done!

> not optimized (the each buffer has unique `ID` and `offset`). The first mode is default, the second one might be used for debugging in cases when

Reviewer 3: ```suggestion
not optimized (the each buffer has unique `ID` and `offset`). The first mode is the default one, while the second one might be used for debugging the optimized version.
```

Author: Done, thanks!

there is possibility incorrect work of the optimized algorithm.
The optimized algorithm `AllocateBuffers` has the main following steps:

> 1. `IdentifyBuffers` - analyzes `Buffers` using graph coloring algorithm to avoid redundant pointer increments by assigning `ID` to them.

Reviewer 3: ```suggestion
1. `IdentifyBuffers` - analyzes `Buffers`  access patterns to avoid redundant pointer increments. A graph coloring algorithm is utilized for this purpose.
```

Author: Great! Thanks!

> 2. `DefineBufferClusters` - creates `BufferClusters` where `Buffers` from the same cluster will have the same `offset` relative to the `Buffer scratchpad` data pointer.

Reviewer 3: It is not clear what `cluster` means in this context, and how these clusters are created.
We already colored our graph, so why do we even need these clusters at all?

Author: Added more details. I hope it's clearer now 😃

The `increment` defines how many data entries are processed on every loop iteration (it usually equals to vector size for the innermost loops of elementwise subgraph). 
So if a loop's `work_amount` is not evenly divisible by its `increment`, it means that a tail processing is required. 
`InsertTailLoop` duplicates the body of such a loop, rescales pointer increments and load/store masks appropriately, and injects these `Ops` immediately after the processed loop.
3. `CleanupLoopOffsets` "fuses" the finalization offsets of loop with an outer loop's pointer increments and zeroes the offsets before `Result` operations.

> 4. `OptimizeLoopSingleEvaluation` moves all pointer arithmetic to finalization offsets in `LoopEnd` and sets the special flag to `LoopEnd` for the next optimizations in the corresponding emitter

Reviewer 3: ```suggestion
4. `OptimizeLoopSingleEvaluation` moves all pointer arithmetic to finalization offsets in `LoopEnd`, and marks the loops that will be executed only once. This information will be used during code emission to eliminate redundant instructions.
```
+ remove linebreak please

Author: It's better. Thanks!
